proposals/0651-ffi-profiling.rst

```diff
@@ -0,0 +1,235 @@
+in haskell code. Currently when trying to profile code like this we get a profile that reports 100% of the time
+spent under ``haskell_work`` and a runtime of merely ~0.5 seconds. Despite the real runtime being over 4 seconds.
+
+.. code-block:: text
+
+    ...
+    total time  =        0.54 secs   (535 ticks @ 1000 us, 1 processor)
+    ...
+
+    COST CENTRE  MODULE SRC               %time %alloc
+
+    c_ffi        Main   Main.hs:8:32-49    88.3    0.0
+    haskell_work Main   Main.hs:13:40-60   11.6  100.0
+
+But if I use my WIP branch of GHC for the same program I get something far closer to reality:
+
+.. code-block:: text
+
+    COST CENTRE  MODULE SRC               %time %alloc
+
+    c_ffi        Main   Main.hs:8:32-49    93.4    0.0
+    haskell_work Main   Main.hs:13:40-60    6.5  100.0
+
+Effect and Interactions
+-----------------------
+The effect of this proposal is two-fold.
+
+The flags and runtime option allow users to quickly get an idea for how long safe ffi
+calls are running (wall time). Allowing them to either rule out or diagnose performance issues as
+having to do with safe fii calls quickly.
+
+The extension of the import spec for ffi calls allows users to selectively keep track of
+time spent on safe ffi calls which are typically expected to be non-blocking but potentially
+long-running and relevant for performance.
+
+However safe ffi calls to blocking functions can still be ignored in time profiles by using the
+default behaviour or marking them as unprofiled explicitly.
+
+Costs and Drawbacks
+-------------------
+While not trivial the changes required for this proposal both in the language specification and
+the GHC implementation are surprisingly small.
+
+For people learning about profiling the proposal should not add much confusion. Safe ffi calls are
+already special and something one needs to know about. Under this proposal this would merely be more
+explicit.
+
+One potential drawback of this proposal is that blocking safe ffi calls might be interpreted as
+taking up system resources if such functions are marked as profiled. Either manually or via one
+of the flags. However as this requires marking a blocking ffi call as safe which is opt-in this
+should be a rare occurrence, far outweighed by the benefit of being able to track longer non-blocking
+safe ffi calls.
+
+The proposal also changes the notion of time used by the profiler slightly. Currently the
+profiler measures only time spent executing code within a haskell context. After this proposal
+safe ffi calls are another potential source of time spent in profiles. This means the total time
+reported is no longer bounded by ``-N * wall_time`` and can now include non-cpu time. In particular
+time tracked will consist of the sum of:
+
+* CPU time spent executing haskell code.
+* CPU time spent executing unsafe ffi calls.
+* CPU time spent executing safe ffi calls.
+* Wall time safe ffi calls are blocked.
```

Comment:
I agree with this. I'd be much happier if we can figure out a way to report separate wall clock and CPU time measurements, rather than conflating them.

In general it's not obvious to me how a library author should choose whether to mark their FFI imports as `profiled` or `unprofiled`. After all, the point of profiling is usually to find out which calls are slow, without necessarily knowing in advance.

Reply:
Given that GHC already counts time spent in unsafe ffi calls as cpu time independent of their blocking behaviour I think it's acceptable to extend that treatment optionally to safe ffi calls which this proposal would allow.

It would no doubt be desirable for GHC and profiling tools to differentiate this sources of time spent further and I expect that to happen eventually. Either in the initial implementation if it is easy to do or at a later point as the feature sees more use. 

But I think differentiating these should be left up to the implementation and not make it into the language specification itself.

I would be happy to add a paragraph that implementations *should* provide a way to distinguish these sources of time spent. I'm not in favor of making this a must requirement.

Reply:
I've added a section `Avoiding confusion from the use of different measures of time:` that outlines this idea.

Reply:
> But I think differentiating these should be left up to the implementation and not make it into the language specification itself.

That's a good point. But then should the proposal be discussing how the time reported by the profiler will look like? As in:

> This means the total time￼reported is no longer bounded by ``-N * wall_time`` and can now include non-cpu time.

I also share this question from @adamgundry:

> In general it's not obvious to me how a library author should choose whether to mark their FFI imports as profiled or unprofiled. After all, the point of profiling is usually to find out which calls are slow, without necessarily knowing in advance.

Looks like the proposal should motivate why FFI bindings need to be tagged in the source code for profiling. Perhaps it is similar to why a user might decide to introduce manual SCCs (AFAIU to prune profiling overheads). This far, the motivation section could be addressed only by adding extra compiler flags for profiling.

Reply:
> Looks like the proposal should motivate why FFI bindings need to be tagged in the source code for profiling. 
> ... This far, the motivation section could be addressed only by adding extra compiler flags for profiling. 

Sensible defaults are the most important thing for a feature like this. Every users dream is that they can run cabal with --enable-profiling, run their executable with +RTS -p, and get a line in big bold letters that let's them know what their performance bottleneck is.

But enabling tracking of *all* safe ffi calls, in particular when tracking runtime instead of cpu time, is bound to generate a lot of false-positives for potential performance bottlenecks. As we can't distinguish functions being bottlenecks from those who are naturally expected to block for a long amount of time. Making this a bad default. Similarly ignoring safe calls also doesn't always make for a great default.

However I expected that authors defining an ffi import will generally know if they expect the ffi call to block for significant amount of time or if including the ffi call in the profile makes sense. Adding a keyword simply gives library and application authors to set sensible profiling defaults on a per-import level. I will try to write this out more clearly in the motivation.

> That's a good point. But then should the proposal be discussing how the time reported by the profiler will look like? 

Personally I always relied on tools like eventlog2html or speedscope for profiling in all but the most basic cases.  
As far as the output of the current `+RTS -p` option is concerned I'm not yet sure what will be best.